p2panda-net/src/engine/engine.rs

```diff
@@ -439,410 +316,71 @@ where
-                    if !topics.has_successfully_joined(&topic.id()).await {
-                        // @TODO: We're dropping messages silently for now, later we want to buffer
-                        // them somewhere until we've joined the topic gossip
-                        continue;
-                    }
-
-                    let result = match event {
-                        ToNetwork::Message { bytes } => {
-                            gossip_actor_tx
-                                .send(ToGossipActor::Broadcast {
-                                    topic_id: topic.id(),
-                                    bytes,
-                                })
-                                .await
-                        }
-                    };
-
-                    if let Err(err) = result {
-                        error!("failed broadcasting message to gossip for topic {topic:?}: {err}");
-                        break;
-                    }
-                }
-            });
-        }
-
-        if self.network_joined {
-            self.announce_topics().await?;
-        }
-
-        Ok(())
-    }
-
-    /// Join all earmarked topics which have not yet been successfully joined.
-    async fn join_earmarked_topics(&mut self) -> Result<()> {
-        for topic_id in self.topics.earmarked().await {
-            if !self.topics.has_successfully_joined(&topic_id).await {
-                self.join_topic(topic_id).await?;
-            }
-        }
-
-        Ok(())
-    }
+        self.topic_streams
+            .subscribe(
+                topic.clone(),
+                from_network_tx,
+                to_network_rx,
+                gossip_ready_tx,
+            )
+            .await?;
 
-    /// Process an message forwarded from the sync actor.
-    async fn on_sync_message(
-        &mut self,
-        header: Vec<u8>,
-        payload: Option<Vec<u8>>,
-        delivered_from: PublicKey,
-        topic_id: [u8; 32],
-    ) -> Result<()> {
-        if self.topics.has_joined(&topic_id).await {
-            self.topics
-                .on_sync_message(topic_id, header, payload, delivered_from)
-                .await?;
-        } else {
-            warn!("received message for unknown topic {topic_id:?}");
-        }
+        // Hot path: Announce our "topics of interest" into the network, hopefully this will speed
+        // up finding other peers.
+        let my_topic_ids = self.topic_streams.topic_ids();
+        self.topic_discovery.announce(my_topic_ids).await?;
 
         Ok(())
     }
 
-    /// Process an inbound message from the network.
+    /// Process an inbound message from one of the gossip overlays.
+    ///
+    /// If the network comes from the "network-wide" gossip overlay (determined by the "network
+    /// id"), then it gets handled by the "topic discovery" mechanism. Otherwise it is a regular,
+    /// custom application-related gossip overlay around a "topic id".
```

Comment:
```suggestion
    /// If the message comes from the "network-wide" gossip overlay (determined by the "network
    /// id"), then it gets handled by the "topic discovery" mechanism. Otherwise it is from a regular,
    /// custom application-related gossip overlay around a "topic id".
```